README.md
SEG Shiny Heatmap (version 0.3.3)
================
Martin Frigaard
2020-04-17

# Welcome to the SEG Shiny app project page\!

<img src="image/seg600-1280-640.png" width="1280" />

Document info:

  - **Created date:** 2020-04-17

  - **R version:** R version 3.6.3 (2020-02-29)

***This is not the code for the working app\!***

This page outlines the code used to develop the SEG application. The
working version is on the Diabetes Technology Society website
[here](https://www.diabetestechnology.org/seg/).

The preview version (not necessarily stable) is available
[here](https://mfrigaard.shinyapps.io/seg-shiny-033/).

For questions, issues, or feature requests, please email Martin at
<support@quesgen.com>.

## Required packages

``` r
library(dplyr) # Data wrangling, glimpse(50) and tbl_df().
library(ggplot2) # Visualise data.
library(lubridate) # Dates and time.
library(readr) # Efficient reading of CSV data.
library(stringr) # String operations.
library(tibble) # Convert row names into a column.
library(tidyr) # Prepare a tidy dataset, gather().
library(magrittr) # Pipes %>%, %T>% and equals(), extract().
library(tidyverse) # all tidyverse packages
library(mosaic) # favstats and other summary functions
library(fs) # file management functions
library(shiny) # apps
library(datapasta) # for pasting tibbles
library(styler) # cleaner code
```

## The Research

This application incorporates two lines of research: the use a
surveillance error grid to evaluate blood glucose measurements, and the
accuracy of marketed blood glucose monitors.

### The Surveillance Error Grid

1.  [The surveillance error grid. Klonoff DC, Lias C, Vigersky R, Clarke
    W, Parkes JL, Sacks DB, Kirkman MS, Kovatchev B; ErrorGrid Panel. J
    Diabetes Sci Technol. 2014
    Jul;8(4):658-72](http://journals.sagepub.com/doi/full/10.1177/1932296814539589).

2.  [Computing the surveillance error grid analysis: procedure and
    examples. Kovatchev BP, Wakeman CA, Breton MD, Kost GJ, Louie RF,
    Tran NK, Klonoff DC. J Diabetes Sci Technol. 2014
    Jul;8(4):673-84.](https://journals.sagepub.com/doi/full/10.1177/1932296814539590)

### Blood Glucose Monitor Surveillance:

1.  [Investigation of the Accuracy of 18 Marketed Blood Glucose
    Monitors. David C. Klonoff, Joan Lee Parkes, Boris P. Kovatchev,
    David Kerr, Wendy C. Bevier, Ronald L. Brazg, Mark Christiansen,
    Timothy S. Bailey, James H. Nichols and Michael A. Kohn. Diabetes
    Care 2018
    Aug; 41(8): 1681-1688.](http://care.diabetesjournals.org/content/41/8/1681.long)

### Feature requests, bugs, etc.

Please email `mjfrigaard@gmail.com` for any bugs or issues.

### CHANGELOG

Any updates and recent changes have been documented
[here](https://github.com/mjfrigaard/seg-shiny-0.3.3/blob/master/CHANGELOG.md).

### Software Requirement Specifications (SRS)

[This
document](https://github.com/mjfrigaard/seg-shiny-0.3.3/blob/master/SRS-documentation.md)
outlines the goals and objectives of the SEG shiny application, and
documents the code in great depth.

<img src="image/quesgen-rstudio-icon-logo.png" width="708" />


CHANGELOG.md
CHANGELOG for SEG Shiny Heatmap (version 0.3.3)
================
Martin Frigaard
2020-04-17

# Welcome to the SEG Shiny app (version 0.3.3)

***This is not the code for the working app\!***

This page outlines the code used to develop the SEG application. The
working version is on the Diabetes Technology Society website
[here](https://www.diabetestechnology.org/seg/).

The preview version (not necessarily stable) is available [here]().

For questions, issues, or feature requests, please email Martin at
<support@quesgen.com>.

## Required packages

``` r
library(dplyr) # Data wrangling, glimpse(50) and tbl_df().
library(ggplot2) # Visualise data.
library(lubridate) # Dates and time.
library(readr) # Efficient reading of CSV data.
library(stringr) # String operations.
library(tibble) # Convert row names into a column.
library(tidyr) # Prepare a tidy dataset, gather().
library(magrittr) # Pipes %>%, %T>% and equals(), extract().
library(tidyverse) # all tidyverse packages
library(mosaic) # favstats and other summary functions
library(fs) # file management functions
library(shiny) # apps
library(datapasta) # for pasting tibbles
library(styler) # cleaner code
```

Document info:

  - **Created date:** 2020-04-17

  - **R version:** R version 3.6.3 (2020-02-29)

## Issues/bugs/fixes

The app was throwing some errors on the DTS site

<img src="image/2020-04-17-seg-error-dts.png" width="1860" />

This looks like a problem in the helpers file, but will investigate and
confirm.

There is also a problem with a SEG table summary, specifically, the
`segTable()` function and it’s interactions with DT.

``` r
source("App/helpers.R")
segTable(dat = base::paste0(github_data_root, 
                            "VanderbiltComplete.csv")) %>% 
  # get data grouped/tallied by risk_cat
        dplyr::count(risk_cat, sort = TRUE) %>%
  # join to lkpSEGRiskCat4 table
        dplyr::full_join(x = .,
                         y = lkpSEGRiskCat4, 
                         by = "risk_cat") %>%
  # create new risk_cat
            dplyr::mutate(
            risk_cat = base::as.numeric(risk_cat),
            # and percent
            Percent = base::paste0(base::round(n / nrow(VandComp) * 100,
                                               digits = 1),
                                   if_else(condition = is.na(n),
                                          true = "", false = "%"))) %>%
  # sort the n column
          dplyr::arrange(desc(n)) %>%
  # rename everything prettier names
          dplyr::select(
                `SEG Risk Level` = risk_cat, # changed to level
                `SEG Risk Category` = risk_cat_txt,
                `Number of Pairs` = n,
                Percent) %>%
 # pass to color formatting
                DT::datatable(., options = list(lengthChange = FALSE,
                                         dom = 't',
                                         rownames = FALSE )) %>%
                # select numerical reference
                DT::formatStyle('SEG Risk Level',
                      target = "row",
                      backgroundColor = DT::styleEqual(
                          levels =  # eight levels/labels
                                c(0, 1, 2, 3,
                                    4, 5, 6, 7),
                            values = c("#00EE00", "#ADFF2F", "#FFFF00",
                                       "#FFD700", "#FFA500","#EE7600",
                                       "#FF4500", "#FF0000")))
```

![](image/helpers.R-1.png)<!-- -->

This was fixed with a new column name for `SEG Risk Level`.

## Error with SEG graph

The SEG tab was throwing this error:

`transformation for secondary axes must be a function`

And this was debugged below. In order to build the plot with the
Gaussian smoothed background, a new image `seg_gaussian_layer_shiny`
needed to be built in the `helpers.R` file

``` r
# 1.0 - DEFINE heatmap inputs ============= 
# 1.0.1 read seg600.png with readPNG ----- -----
library(jpeg)
library(png)
# load the seg600.png file
download.file(url = "https://raw.githubusercontent.com/mjfrigaard/seg-shiny-data/master/Image/seg600.png", 
              destfile = "Image/seg600.png")
BackgroundSmooth <- png::readPNG("Image/seg600.png")
# 1.0.2 mmol conversion factor ---- -----
mmolConvFactor <- 18.01806
# 1.0.3 rgb2hex function ---- -----
# This is the RGB to Hex number function for R
rgb2hex <- function(r, g, b) rgb(r, g, b, maxColorValue = 255)
# 1.0.4 risk factor colors ----  -----
# These are the values for the colors in the heatmap.
abs_risk_0.0000_color <- rgb2hex(0, 165, 0)
# abs_risk_0.0000_color
abs_risk_0.4375_color <- rgb2hex(0, 255, 0)
# abs_risk_0.4375_color
abs_risk_1.0625_color <- rgb2hex(255, 255, 0)
# abs_risk_1.0625_color
abs_risk_2.7500_color <- rgb2hex(255, 0, 0)
# abs_risk_2.7500_color
abs_risk_4.0000_color <- rgb2hex(128, 0, 0)
# abs_risk_4.0000_color
riskfactor_colors <- c(
  abs_risk_0.0000_color,
  abs_risk_0.4375_color,
  abs_risk_1.0625_color,
  abs_risk_2.7500_color,
  abs_risk_4.0000_color
)
# 1.0.5 create base_data data frame ---- ----- 
base_data <- data.frame(
  x_coordinate = 0,
  y_coordinate = 0,
  color_gradient = c(0:4)
)
# These are the data objects used to define various components in the graph. 

# First create the base layer, but add the axes formats and adjust the point 
# to be almost non-existent. 
# - create base_layer plot ---- 
base_layer <- base_data %>% 
  ggplot(aes(
      x = x_coordinate,
      y = y_coordinate,
      fill = color_gradient)) +
  geom_point(size = 0.00000001,
             color = "white")
# scales_layer ------
scales_layer <- base_layer + 
  ggplot2::scale_y_continuous(
    limits = c(0, 600),
    sec.axis =
      sec_axis(~. / mmolConvFactor,
        name = "Measured blood glucose (mmol/L)"
      ),
    name = "Measured blood glucose (mg/dL)"
  ) +
  scale_x_continuous(
    limits = c(0, 600),
    sec.axis =
      sec_axis(~. / mmolConvFactor,
        name = "Reference blood glucose (mmol/L)"
      ),

App/helpers.R
# =====================================================================#
# This is the helper code to create: data inputs, mmol constant, colors
# plot canvas data, and a function for cleaning/preparing the uploaded
# data files to the shiny app.
# Authored by and feedback to mjfrigaard@gmail.com
# MIT License
# Version: 0.1.3.3 ----
# =====================================================================#

options(shiny.maxRequestSize = 30 * 1024^2)
library(colourpicker)
library(shinythemes)
library(tidyverse)
library(styler)
library(heatmaply)
library(tools) # for tools
library(data.table) # for fread() function
library(datapasta)
library(clipr)
require(dplyr) # Data wrangling, glimpse(50) and tbl_df().
require(ggplot2) # Visualise data.
require(lubridate) # Dates and time.
require(readr) # Efficient reading of CSV data.
require(stringr) # String operations.
require(tibble) # Convert row names into a column.
require(tidyr) # Prepare a tidy dataset, gather().
require(magrittr) # Pipes %>%, %T>% and equals(), extract().
require(mosaic) # favstats and other summary functions
require(fs) # file management functions
require(shiny) # apps
require(datapasta) # for pasting tibbles

# 0.0.0 - DEFINE data inputs ============= ----


# 0.0.1 - define github data repo ---- ---- 
github_data_root <- "https://raw.githubusercontent.com/mjfrigaard/seg-shiny-data/master/Data/"

# 0.0.2 - VanderbiltComplete file ---- ---- 
full_sample_repo <- base::paste0(github_data_root, 
                                 "VanderbiltComplete.csv")

# 0.0.3 - the AppRiskPairData file  ---- ---- 
app_riskpair_repo <- base::paste0(github_data_root, 
                                  "AppRiskPairData.csv")
# app_riskpair_repo
# 0.0.4 - the AppLookUpRiskCat file  ---- ---- 
app_lookup_repo <- base::paste0(github_data_root, 
                                "AppLookUpRiskCat.csv")

# 0.0.5 - the APPSEGBlandAltmanRefVals file  ---- ---- 
ba_ref_repo <- base::paste0(github_data_root, 
                           "APPSEGBlandAltmanRefVals.csv")

# # 0.1.0 - IMPORT DATA ============= ----

# this is not needed for the app, but loaded for the README.Rmd
VandComp <- readr::read_csv(full_sample_repo )

# 0.1.1 - load RiskPairData ----
RiskPairData <- readr::read_csv(file = app_riskpair_repo)

# 0.1.2 - LookUpRiskCat ----
LookUpRiskCat <- readr::read_csv(file = app_lookup_repo)

# 0.1.3 - SEGBlandAltmanRefVals -----
SEGBlandAltmanRefVals <- readr::read_csv(file = ba_ref_repo)

# 0.1.4 - define risk grade look-up table lkpRiskGrade  ---- ---- ----
# for RiskGradeTable3
lkpRiskGrade <- tibble::tribble(
~`risk_grade_id`, ~`risk_grade`, ~`REF`,
1, "A", "0 - 0.5",
2, "B", "> 0.5 - 1.0",
3, "C", "> 1.0 - 2.0",
4, "D", "> 2.0 - 3.0",
5, "E", "> 3.0")
lkpRiskGrade

# 0.1.5 - define lkpSEGRiskCat4 ---- ---- ---- 
lkpSEGRiskCat4 <- tibble::tribble(
    ~risk_cat,      ~risk_cat_txt, ~ABSLB, ~ABSUB,
    0L,             "None", -0.001,    0.5,
    1L,    "Slight, Lower",    0.5,      1,
    2L,   "Slight, Higher",      1,    1.5,
    3L,  "Moderate, Lower",    1.5,      2,
    4L, "Moderate, Higher",      2,    2.5,
    5L,    "Severe, Lower",    2.5,      3,
    6L,    "Severe, Upper",      3,    3.5,
    7L,          "Extreme",      3,   1000
)

# lkpSEGRiskCat4$risk_cat <- as.numeric(lkpSEGRiskCat4$risk_cat)

# 0.1.6 - define lkpISORanges look-up table for ISORangeTable5  ---- ---- ----
lkpISORanges <- tibble::tribble(
       ~ID,                ~iso_range,
        1L,    "<= 5% or 5 mg/dL",
        2L,  "> 5 - 10% or mg/dL",
        3L, "> 10 - 15% or mg/dL",
        4L,    "> 15 - 20% mg/dL",
        5L,   "> 20% or 20 mg/dL")
# lkpISORanges

# 1.0 - DEFINE heatmap inputs ============= ----
library(jpeg)
library(png)

# 1.0.0 - create base_data data frame ---- ----- 
base_data <- data.frame(
  x_coordinate = 0,
  y_coordinate = 0,
  color_gradient = c(0:4)
)

# 1.0.1 read seg600.png with readPNG ----- -----
# load the seg600.png file
download.file(url = "https://raw.githubusercontent.com/mjfrigaard/seg-shiny-data/master/Image/seg600.png", 
              destfile = "seg600.png")
BackgroundSmooth <- png::readPNG("seg600.png")
# 1.0.2 mmol conversion factor ---- -----
mmolConvFactor <- 18.01806
# 1.0.3 rgb2hex function ---- -----
# This is the RGB to Hex number function for R
rgb2hex <- function(r, g, b) rgb(r, g, b, maxColorValue = 255)
# 1.0.4 risk factor colors ----  -----
# These are the values for the colors in the heatmap.
abs_risk_0.0000_color <- rgb2hex(0, 165, 0)
# abs_risk_0.0000_color
abs_risk_0.4375_color <- rgb2hex(0, 255, 0)
# abs_risk_0.4375_color
abs_risk_1.0625_color <- rgb2hex(255, 255, 0)
# abs_risk_1.0625_color
abs_risk_2.7500_color <- rgb2hex(255, 0, 0)
# abs_risk_2.7500_color
abs_risk_4.0000_color <- rgb2hex(128, 0, 0)
# abs_risk_4.0000_color
riskfactor_colors <- c(
  abs_risk_0.0000_color,
  abs_risk_0.4375_color,
  abs_risk_1.0625_color,
  abs_risk_2.7500_color,
  abs_risk_4.0000_color
)

# 1.0.5 - create scales_layer plot ---- 
base_layer <- base_data %>% 
  ggplot(aes(
    x = x_coordinate,
    y = y_coordinate,
    fill = color_gradient)) +
  geom_point(size = 0.00000001,
             color = "white")
scales_layer <- base_layer + 
  ggplot2::scale_y_continuous(
    limits = c(0, 600),
    sec.axis =
      sec_axis(~. / mmolConvFactor,
               name = "Measured blood glucose (mmol/L)"
      ),
    name = "Measured blood glucose (mg/dL)"
  ) +
  scale_x_continuous(
    limits = c(0, 600),
    sec.axis =
      sec_axis(~. / mmolConvFactor,
               name = "Reference blood glucose (mmol/L)"
      ),
    name = "Reference blood glucose (mg/dL)"
  )
# scales_layer
# 1.0.6 - add gaussian_layer ---- 
gaussian_layer <- scales_layer +
  ggplot2::annotation_custom(
    grid::rasterGrob(image = BackgroundSmooth, 
                     width = unit(1,"npc"), 
                     height = unit(1,"npc")), 
    xmin = 0, 
    xmax =  600, 
    ymin = 0, 
    ymax =  600) 
# gaussian_layer

# 1.0.7 - create seg_gaussian_layer_shiny ----
seg_gaussian_layer_shiny <- gaussian_layer + 
  ggplot2::scale_fill_gradientn( # scale_*_gradientn creats a n-color gradient
    values = scales::rescale(c(
      0, # darkgreen
      0.4375, # green
      1.0625, # yellow
      2.75, # red
      4.0 # brown
    )), 
    limits = c(0, 4),
    colors = riskfactor_colors,
    guide = guide_colorbar(
      ticks = FALSE,
      barheight = unit(100, "mm")
    ),
    breaks = c(
      0.25,
      1,
      2,
      3,
      3.75
    ),
    labels = c(
      "none",
      "slight",
      "moderate",
      "high",
      "extreme"
    ),
    name = "risk level")

# 2.0 - DEFINE pairtypeTable function ============= ----
pairtypeTable <- function(dat) {
  # 2.1.1 - import data frame/define SampMeasData ----
  SampMeasData <- readr::read_csv(dat)
  # 2.1.2 - convert the columns to numeric  ----
  SampMeasData <- SampMeasData %>%
    dplyr::mutate(BGM = as.double(BGM),
                  REF = as.double(REF)) %>%
    # 2.2 create bgm_pair_cat ---- ----  ----
  dplyr::mutate(
    bgm_pair_cat =
      dplyr::case_when(
        BGM < REF ~ "BGM < REF",
        BGM == REF ~ "BGM = REF",
        BGM > REF ~ "BGM > REF"
      )
  ) %>%
    # 2.3 create excluded ---- ----  ----
  dplyr::mutate(
    excluded =
      dplyr::case_when(
        REF > 600 ~ "REF > 600: Excluded from SEG Analysis",
        TRUE ~ NA_character_
      )
  ) %>%
    # 2.4 create included ---- ----  ----
  dplyr::mutate(
    included =
      dplyr::case_when(
        REF <= 600 ~ "Total included in SEG Analysis",
        REF > 600 ~ "Total excluded in SEG Analysis"
      )
  )
  # 2.5 create BGMPairs ---- ---- ----
  BGMPairs <- SampMeasData %>%
    dplyr::count(bgm_pair_cat) %>%
    dplyr::rename(
      `Pair Type` = bgm_pair_cat,
      Count = n
    )
  # 2.6 create Excluded ---- ---- ----
  Excluded <- SampMeasData %>%
    dplyr::count(excluded) %>%
    dplyr::rename(
      `Pair Type` = excluded,
      Count = n
    ) %>%
    dplyr::filter(!is.na(`Pair Type`))
  # 2.7 create Included ---- ---- ----
  Included <- SampMeasData %>%
    dplyr::count(included) %>%
    dplyr::rename(
      `Pair Type` = included,
      Count = n
    ) %>%
    dplyr::filter(`Pair Type` == "Total included in SEG Analysis")
  # 2.8 create PairTypeTable ----  ----  ----
  PairTypeTable <- dplyr::bind_rows(BGMPairs, 
                                    Excluded, 
                                    Included)
  # 2.9 add the Total row  ----  ----  ----
  PairTypeTable <- PairTypeTable %>% tibble::add_row(
    `Pair Type` = "Total",
    Count = nrow(SampMeasData),
    .after = 0
  )
  return(PairTypeTable)
}

# PairTypeTest <- pairtypeTable(paste0(github_root, full_sample_repo))
# PairTypeTest

# 3.0 segTable() FUNCTION  ---- ----  ---- ----  ---- ----  ---- ----
segTable <- function(dat) {

  # 3.1 - import data frame ---- ---- -----
  SampMeasData <- suppressWarnings(readr::read_csv(file = dat))

  SampMeasData %>%
    dplyr::mutate(BGM = as.double(BGM),
                  REF = as.double(REF)) %>%

  # 3.2 create bgm_pair_cat ---- ----  ----
    dplyr::mutate(
      bgm_pair_cat =
        dplyr::case_when(
          BGM < REF ~ "BGM < REF",
          BGM == REF ~ "BGM = REF",
          BGM > REF ~ "BGM > REF"
        )
    ) %>%
    # 3.3 create ref_pair_2cat ---- ----  ----
    dplyr::mutate(
      ref_pair_2cat =
        dplyr::case_when(
          REF > 600 ~ "REF > 600: Excluded from SEG Analysis",
          REF < 21 & REF <= 600 ~ "REF <21: Included in SEG Analysis"
        )
    ) %>%
    # # 3.4 create included ---- ----  ----
    dplyr::mutate(
      included =
        dplyr::case_when(
          REF <= 600 ~ "Total included in SEG Analysis",
          REF > 600 ~ "Total excluded in SEG Analysis"
        )
    ) %>%
    # 3.5 join to RiskPairData ---- ----  ----
    dplyr::inner_join(.,
      y = RiskPairData,
      by = c("BGM", "REF")
    ) %>%
    dplyr::mutate( # 3.6 Create risk_cat variable ---- ---- ----
      risk_cat =
        base::findInterval(
          x = abs_risk, # the abs_risk absolute value
          vec = LookUpRiskCat$ABSLB, # the lower bound absolute risk
          left.open = TRUE
        ) - 1
    ) %>%
    dplyr::inner_join( # 3.7 Join to LookUpRiskCat data ---- ----  ----
      x = ., y = LookUpRiskCat, # inner join to look-up
      by = "risk_cat"
    ) %>%
    dplyr::mutate(
      risk_cat_txt = # text risk categories
      dplyr::case_when(
        abs_risk < 0.5 ~ "None",
        abs_risk >= 0.5 & abs_risk <= 1 ~ "Slight, Lower",
        abs_risk > 1 & abs_risk <= 1.5 ~ "Slight, Higher",
        abs_risk > 1.5 & abs_risk <= 2.0 ~ "Moderate, Lower",
        abs_risk > 2 & abs_risk <= 2.5 ~ "Moderate, Higher",
        abs_risk > 2.5 & abs_risk <= 3.0 ~ "Severe, Lower",
        abs_risk > 3.0 & abs_risk <= 3.5 ~ "Severe, Higher",
        abs_risk > 3.5 ~ "Extreme"
      )
    ) %>%
    dplyr::mutate(
      rel_diff = (BGM - REF) / REF, # relative diff
      abs_rel_diff = abs(rel_diff), # abs relative diff
      sq_rel_diff = rel_diff^2,
      iso_diff =
      if_else(REF >= 100, # condition 1
        100 * abs(BGM - REF) / REF, # T 1
        if_else(REF < 100, # condition 2
          abs(BGM - REF), # T 2
          NA_real_
        ), # F 2
        NA_real_
      ), # F1
      iso_range = # # 4.3.16 create iso range variable ----
      dplyr::case_when(
          # # A tibble: 5 x 2
          #      ID iso_range
          #   <int> <chr>
          # 1     1 <= 5% or 5 mg/dL
          # 2     2 > 5 - 10% or mg/dL
          # 3     3 > 10 - 15% or mg/dL
          # 4     4 > 15 - 20% mg/dL
          # 5     5 > 20% or 20 mg/dL
        iso_diff <= 5 ~ "<= 5% or 5 mg/dL",
        iso_diff > 5 & iso_diff <= 10 ~ "> 5 - 10% or mg/dL",
        iso_diff > 10 & iso_diff <= 15 ~ "> 10 - 15% or mg/dL",
        iso_diff > 15 & iso_diff <= 20 ~ "> 15 - 20% mg/dL",
        iso_diff > 20 ~ "> 20% or 20 mg/dL"),

      risk_grade = dplyr::case_when(
        abs_risk >= 0.0 & abs_risk < 0.5 ~ "A",
        abs_risk >= 0.5 & abs_risk < 1.0 ~ "B",
        abs_risk >= 1.0 & abs_risk < 2.0 ~ "C",
        abs_risk >= 2.0 & abs_risk < 3.0 ~ "D",
        abs_risk >= 3.0 ~ "E"
      ),
      risk_grade_txt = dplyr::case_when(
        abs_risk >= 0.0 & abs_risk < 0.5 ~ "0 - 0.5",
        abs_risk >= 0.5 & abs_risk < 1.0 ~ "> 0.5 - 1.0",
        abs_risk >= 1.0 & abs_risk < 2.0 ~ "> 1.0 - 2.0",
        abs_risk >= 2.0 & abs_risk < 3.0 ~ "> 2.0 - 3.0",
        abs_risk >= 3.0 ~ "> 3.0"
      )
    )
}

# test segTable
# TestSEGTableFunc <- segTable(paste0(github_root, full_sample_repo))
# TestSEGTableFunc %>% dplyr::count(included)

# 4.5.0 binomialTable ---- ----  ---- ----  ---- ----  ---- ----
binomialTable <- function(dat) {

  dataset <- segTable(dat)

  CompliantPairs <- nrow(dataset) - base::nrow(dplyr::filter(
    dataset, iso_diff > 15)) %>%
  tibble(
    `Compliant Pairs` = .)
# CompliantPairs
# Then calculate the percent
CompliantPairs <- CompliantPairs %>%
  dplyr::mutate(
    `Compliant Pairs %` =
      base::paste0(base::round(
        100 * `Compliant Pairs` / nrow(dataset),
        1
      ), "%") )
# CompliantPairs
# create probability
prb <- 0.95
p_value <- 0.05
df_size <- nrow(dataset)
qbinom_tibble <- qbinom(
    p = p_value,
    size = df_size,
    prob = prb) %>%
    tibble(`value` = .) %>%
    # clean up this variable in the tibble for display
    dplyr::rename(`Lower Bound for Acceptance` = value)
# qbinom_tibble
QbinomTable <- qbinom_tibble %>%
  dplyr::mutate(
    `Lower Bound for Acceptance %` =
      base::paste0(base::round(
        100 * `Lower Bound for Acceptance` / nrow(dataset),
        1
      ), "%")
  )
BinomialTest6 <- bind_cols(CompliantPairs, QbinomTable)
BinomialTest6 <- BinomialTest6 %>% dplyr::mutate(
  Result =
    if_else(condition = `Compliant Pairs` < `Lower Bound for Acceptance`,
      true = paste0(
        BinomialTest6$`Compliant Pairs %`[1],
        " < ",
        BinomialTest6$`Lower Bound for Acceptance %`[1],
        " - Does not meet BGM Surveillance Study Accuracy Standard"
      ),
      false = paste0(
        BinomialTest6$`Compliant Pairs %`[1],
        " > ",
        BinomialTest6$`Lower Bound for Acceptance %`[1],
        " - Meets BGM Surveillance Study Accuracy Standard"
      )
    )
)
return(BinomialTest6)
}

# test binomialTable
# TestBinomial <- binomialTable(paste0(github_root, full_sample_repo))

